packages/workbox-routing/src/lib/router.js

```diff
@@ -46,113 +51,152 @@ import normalizeHandler from './normalize-handler';
+   * if there is a match.
+   *
+   * @example
+   * const imageRoute = new RegExpRoute({
+   *   regExp: /images/,
+   *   handler: new CacheFirst(),
+   * });
+   *
+   * const router = new Router();
+   * router.registerRoute({route: imageRoute});
+   * router.addFetchListener();
+   */
+  addFetchListener() {
+    self.addEventListener('fetch', (event) => {
+      const responsePromise = this.handleRequest({event});
+      if (responsePromise) {
+        event.respondWith(responsePromise);
+      }
+    });
+  }
+
+  /**
+   * This can be used to apply the routing rules to generate a response for a
+   * given request inside your own `fetch` event handler.
+   *
+   * @example
+   * const imageRoute = new RegExpRoute({
+   *   regExp: /images/,
+   *   handler: new CacheFirst(),
+   * });
+   *
+   * const router = new Router();
+   * router.registerRoute({route: imageRoute});
+   *
+   * self.addEventListener('fetch', (event) => {
+   *   event.waitUntil((async () => {
+   *     let response = await router.handleRequest({event});
+   *     // Do something with response, and then eventually respond with it.
+   *     event.respondWith(response);
+   *   })());
+   * });
+   *
+   * @param {Object} input
+   * @param {FetchEvent} input.event The event passed in to a `fetch` handler.
+   * @return {Promise<Response>|undefined} Returns a promise for a response,
+   * taking the registered routes into account. If there was no matching route
+   * and there's no `defaultHandler`, then returns undefined.
+   */
+  handleRequest({event}) {
+    isInstance({event}, FetchEvent);
+    const url = new URL(event.request.url);
+    if (!url.protocol.startsWith('http')) {
+      logHelper.log({
+        that: this,
+        message: `The URL does not start with HTTP, so it can't be handled.`,
+        data: {
+          request: event.request,
+        },
+      });
+      return;
+    }
 
-    if (handleFetch) {
-      this._addFetchListener();
+    let {handler, params} = this._findHandlerAndParams({event, url});
+
+    // If we don't have a handler because there was no matching route, then
+    // fall back to defaultHandler if that's defined.
+    if (!handler && this.defaultHandler) {
+      handler = this.defaultHandler;
+    }
+
+    if (handler) {
+      let responsePromise = handler.handle({url, event, params});
+      if (this.catchHandler) {
+        responsePromise = responsePromise.catch((error) => {
+          return this.catchHandler.handle({url, event, error});
+        });
+      }
+      return responsePromise;
     }
   }
 
   /**
-   * This method will actually add the fetch event listener.
+   * Checks the incoming even.request against the registered routes, and if
+   * there's a match, returns the corresponding handler along with any params
+   * generated by the match.
+   *
+   * @param {FetchEvent} input.event
+   * @param {URL} input.url
+   * @return {Object} Returns an object with `handler` and `params` properties
```

Comment:
👍 Matt's suggestion.  If I might suggest one further edit "Returns an object with `handler` and `params `properties. They are populated if a matching route was found or `undefined` otherwise".